File: daily/2020-01-04/33 Search in Rotated Sorted Array.py
class Solution(object):
    def search(self, nums, target):
        """
        :type nums: List[int]
        :type target: int
        :rtype: int
        """
        l = 0
        r = len(nums) - 1
        while l <= r:
            m = (l + r) // 2
            if target == nums[m]:
                return m
            if nums[m] >= nums[0] and (nums[0] <= target < nums[m]):
                r = m - 1
            elif nums[m] >= nums[0]:
                l = m + 1
            elif nums[m] < nums[0] and (nums[m] < target <= nums[-1]):
                l = m + 1
            else:
                r = m - 1
        return -1

File: Leetcode First 200 JavaScript solution/33. Search in Rotated Sorted Array.js
var search = function(nums, target) {
    // binary search
    

    let left = 0, right = nums.length-1;
    const n = nums.length-1
    while (left <= right) {
        
        const mid = left + Math.floor((right - left) /2);
        
        if (nums[mid] === target) return mid;
        
        // left side
        else if (nums[mid] >= nums[0]) {
            if (nums[mid] > target && target >= nums[0]) right = mid - 1;
            else left = mid + 1;
        } 
        else {
            if (nums[mid] < target && target <= nums[n]) left = mid + 1;
            else right = mid - 1;
        }
    }
    return -1;
};

// divid into 3 parts ! here!!
